kamaki/cli/cmds/cyclades.py

```diff
@@ -1145,3 +1148,152 @@ def main(self, server_id):
+
+    def main(self, key_name):
+        super(self.__class__, self)._run()
+        self._run(key_name=key_name)
+
+
+@command(keypair_cmds)
+class keypair_delete(_CycladesInit):
+    """Delete a keypair"""
+
+    @errors.Generic.all
+    @errors.Keypair.connection
+    @errors.Keypair.name
+    def _run(self, key_name):
+        keypair = self.client.delete_keypair(key_name)
+
+    def main(self, key_name):
+        super(self.__class__, self)._run()
+        self._run(key_name=key_name)
+
+
+@command(keypair_cmds)
+class keypair_upload(_CycladesInit, OptionalOutput):
+    """Upload or update a keypair"""
+
+    arguments = dict(
+        key_name=ValueArgument(
+            'The name of the key',
+            '--key-name'),
+        public_key=ValueArgument(
+            'The contents of the keypair(the public key)',
+            '--public-key'),
+        force=FlagArgument(
+            'Force key name override', ('-f', '--force')
+        )
+    )
+
+    required = ['public_key']
+
+    @errors.Generic.all
+    @errors.Keypair.connection
+    def _run(self):
+
+        if self['key_name'] is not None:
+            active_keypairs = self.client.list_keypairs()
+            active_keypairs = [k['keypair'] for k in active_keypairs]
+            if (len(filter(lambda k: k['name'].strip() == self['key_name'],
+                           active_keypairs)) > 0 and not self['force']):
+                raise CLIError(
+                    'Keypair with that name alread exists',
+                    details=['A keypair with the name %s already exists'
+                             % (self['key_name']),
+                             'In order to create a new key, the name of '
+                             'the key must be unique',
+                             'If you want the system to generate a unique '
+                             'name for you leave the --key-name option empty'])
+            key_name = self['key_name']
+        else:
+            uniq = str(uuid.uuid4())[:8]
+            key_name = 'kamaki-key_autogen_{:%b_%d_%H_%M_%S_%f}_{uniq}'.format(
+                datetime.now(), uniq=uniq)
+
+        keypair = self.client.create_or_update_key(
+            key_name=key_name, public_key=self['public_key'])
+        self.print_(keypair, self.print_dict)
+
+    def main(self):
+        super(self.__class__, self)._run()
+        self._run()
+
+
+@command(keypair_cmds)
+class keypair_generate(_CycladesInit, OptionalOutput):
+    """Generate a keypair"""
+
+    arguments = dict(
+        key_name=ValueArgument(
+            'The name of the key',
+            '--key-name'),
+        force=FlagArgument(
+            'Force key name override', ('-f', '--force')
+        )
+    )
+
+    @errors.Generic.all
+    @errors.Keypair.connection
+    def _run(self):
+
+        if self['key_name'] is not None:
+            active_keypairs = self.client.list_keypairs()
+            active_keypairs = [k['keypair'] for k in active_keypairs]
```

Comment:
I don't see where this list is used, except from checking the key names. So, maybe, instead of extracting keypairs, you could extract key names?
```python
active_keypairs = [k['keypair'] for k in self.client.list_keypairs()]
active_keynames = [kp['name'] for kp in active_keypairs]
```